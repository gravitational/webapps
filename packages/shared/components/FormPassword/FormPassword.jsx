/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import PropTypes from 'prop-types';
import { Card, ButtonPrimary } from 'design';
import * as Alerts from 'design/Alert';
import useAttempt from './../../hooks/useAttempt';
import FieldInput from './../FieldInput';
import Validation from '../Validation';
import {
  requiredToken,
  requiredPassword,
  requiredField,
  requiredConfirmedPassword,
} from './../Validation/rules';
import { isOtp, isU2f } from './../../services/enums';

function FormPassword(props) {
  const { onChangePassWithU2f, onChangePass, auth2faType } = props;
  const [attempt, attemptActions] = useAttempt();
  const [token, setToken] = React.useState('');
  const [oldPass, setOldPass] = React.useState('');
  const [newPass, setNewPass] = React.useState('');
  const [newPassConfirmed, setNewPassConfirmed] = React.useState('');

  const otpEnabled = isOtp(auth2faType);
  const u2fEnabled = isU2f(auth2faType);
  const { isProcessing } = attempt;

  function submit() {
    if (u2fEnabled) {
      return onChangePassWithU2f(oldPass, newPass);
    }

    return onChangePass(oldPass, newPass, token);
  }

  function resetForm() {
    setOldPass('');
    setNewPass('');
    setNewPassConfirmed('');
    setToken('');
  }

  function onSubmit(e, validator) {
    e.preventDefault();
    if (!validator.validate()) {
      return;
    }

    validator.reset();

    attemptActions.start();
    submit()
      .then(() => {
        attemptActions.stop();
        resetForm();
      })
      .catch(err => {
        attemptActions.error(err);
      });
  }

  return (
    <Validation>
      {({ validator }) => (
        <Card as="form" bg="primary.light" width="456px" p="6">
          <Status u2f={u2fEnabled} attempt={attempt} />
          <FieldInput
            rule={requiredField('Current Password is required')}
            label="Current Password"
            value={oldPass}
            onChange={e => setOldPass(e.target.value)}
            type="password"
            placeholder="Password"
          />
          {otpEnabled && (
            <FieldInput
              label="2nd factor token"
              rule={requiredToken}
              width="50%"
              value={token}
              onChange={e => setToken(e.target.value)}
              type="text"
              placeholder="OTP Token"
            />
          )}
          <FieldInput
            rule={requiredPassword}
            label="New Password"
            value={newPass}
            onChange={e => setNewPass(e.target.value)}
            type="password"
            placeholder="New Password"
          />
          <FieldInput
            rule={requiredConfirmedPassword(newPass)}
            label="Confirm Password"
            value={newPassConfirmed}
            onChange={e => setNewPassConfirmed(e.target.value)}
            type="password"
            placeholder="Confirm Password"
          />
          <ButtonPrimary
            block
            disabled={isProcessing}
            size="large"
            onClick={e => onSubmit(e, validator)}
            mt={5}
          >
            Update Password
          </ButtonPrimary>
        </Card>
      )}
    </Validation>
  );
}

FormPassword.propTypes = {
  onChangePass: PropTypes.func.isRequired,
  onChangePassWithU2f: PropTypes.func.isRequired,

  /**
   * auth2faType defines login type.
   * eg: u2f, otp, off (disabled).
   *
   * enums are defined in shared/services/enums.js
   */
  auth2faType: PropTypes.string,
};

function Status({ attempt, u2f }) {
  const { isProcessing, isFailed, message, isSuccess } = attempt;

  if (isFailed) {
    return <Alerts.Danger>{message}</Alerts.Danger>;
  }

  if (isSuccess) {
    return <Alerts.Success>Your password has been changed!</Alerts.Success>;
  }

  const waitForU2fKeyResponse = isProcessing && u2f;
  if (waitForU2fKeyResponse) {
    return (
      <Alerts.Info>
        Insert your U2F key and press the button on the key
      </Alerts.Info>
    );
  }

  return null;
}

export default FormPassword;
