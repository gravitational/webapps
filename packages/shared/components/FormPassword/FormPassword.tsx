/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Card, ButtonPrimary, Flex, Box } from 'design';
import * as Alerts from 'design/Alert';
import useAttempt from 'shared/hooks/useAttempt';
import { Option } from 'shared/components/Select';
import FieldInput from '../FieldInput';
import FieldSelect from '../FieldSelect';
import Validation, { Validator } from '../Validation';
import {
  requiredToken,
  requiredPassword,
  requiredField,
  requiredConfirmedPassword,
} from '../Validation/rules';
import { Auth2faType } from 'shared/services';

function FormPassword(props: Props) {
  const { onChangePassWithU2f, onChangePass, auth2faType = 'off' } = props;
  const mfaEnabled = auth2faType === 'on' || auth2faType === 'optional';
  const otpEnabled = auth2faType === 'otp';
  const u2fEnabled = auth2faType === 'u2f';

  const [attempt, attemptActions] = useAttempt({});
  const [token, setToken] = React.useState('');
  const [oldPass, setOldPass] = React.useState('');
  const [newPass, setNewPass] = React.useState('');
  const [newPassConfirmed, setNewPassConfirmed] = React.useState('');
  const [mfaOptions] = React.useState(() => {
    let mfaOptions = [
      { value: 'u2f', label: 'U2F' },
      { value: 'otp', label: 'TOTP' },
    ];
    if (auth2faType === 'optional') {
      mfaOptions = [{ value: 'none', label: 'NONE' }, ...mfaOptions];
    }
    return mfaOptions;
  });
  const [mfaType, setMfaType] = React.useState(mfaOptions[0]);

  const { isProcessing } = attempt;
  const isU2fSelected = u2fEnabled || (mfaEnabled && mfaType.value === 'u2f');

  function submit() {
    if (isU2fSelected) {
      return onChangePassWithU2f(oldPass, newPass);
    }

    return onChangePass(oldPass, newPass, token);
  }

  function resetForm() {
    setOldPass('');
    setNewPass('');
    setNewPassConfirmed('');
    setToken('');
  }

  function onSubmit(
    e: React.MouseEvent<HTMLButtonElement>,
    validator: Validator
  ) {
    e.preventDefault();
    if (!validator.validate()) {
      return;
    }

    validator.reset();

    attemptActions.start();
    submit()
      .then(() => {
        attemptActions.stop();
        resetForm();
      })
      .catch(err => {
        attemptActions.error(err);
      });
  }

  function onSetMfaOption(option: Option, validator: Validator) {
    setToken('');
    attemptActions.clear();
    validator.reset();
    setMfaType(option);
  }

  return (
    <Validation>
      {({ validator }) => (
        <Card as="form" bg="primary.light" width="456px" p="6">
          <Status isU2F={isU2fSelected} attempt={attempt} />
          <FieldInput
            rule={requiredField('Current Password is required')}
            label="Current Password"
            value={oldPass}
            onChange={e => setOldPass(e.target.value)}
            type="password"
            placeholder="Password"
          />
          {mfaEnabled && (
            <Flex alignItems="flex-end" mb={4}>
              <Box width="50%" data-testid="mfa-select">
                <FieldSelect
                  label="Second factor"
                  value={mfaType}
                  options={mfaOptions}
                  onChange={opt => onSetMfaOption(opt as Option, validator)}
                  mr={3}
                  mb={0}
                  isDisabled={isProcessing}
                />
              </Box>
              <Box width="50%">
                {mfaType.value === 'otp' && (
                  <FieldInput
                    label="Two factor token"
                    rule={requiredToken}
                    autoComplete="off"
                    value={token}
                    onChange={e => setToken(e.target.value)}
                    placeholder="123 456"
                    mb={0}
                  />
                )}
              </Box>
            </Flex>
          )}
          {otpEnabled && (
            <FieldInput
              label="2nd factor token"
              rule={requiredToken}
              width="50%"
              value={token}
              onChange={e => setToken(e.target.value)}
              type="text"
              placeholder="OTP Token"
            />
          )}
          <FieldInput
            rule={requiredPassword}
            label="New Password"
            value={newPass}
            onChange={e => setNewPass(e.target.value)}
            type="password"
            placeholder="New Password"
          />
          <FieldInput
            rule={requiredConfirmedPassword(newPass)}
            label="Confirm Password"
            value={newPassConfirmed}
            onChange={e => setNewPassConfirmed(e.target.value)}
            type="password"
            placeholder="Confirm Password"
          />
          <ButtonPrimary
            block
            disabled={isProcessing}
            size="large"
            onClick={e => onSubmit(e, validator)}
            mt={5}
          >
            Update Password
          </ButtonPrimary>
        </Card>
      )}
    </Validation>
  );
}

function Status({ attempt, isU2F }: StatusProps) {
  if (attempt.isFailed) {
    return <Alerts.Danger>{attempt.message}</Alerts.Danger>;
  }

  if (attempt.isSuccess) {
    return <Alerts.Success>Your password has been changed!</Alerts.Success>;
  }

  const waitForU2fKeyResponse = attempt.isProcessing && isU2F;
  if (waitForU2fKeyResponse) {
    return (
      <Alerts.Info>
        Insert your U2F key and press the button on the key
      </Alerts.Info>
    );
  }

  return null;
}

type StatusProps = {
  attempt: ReturnType<typeof useAttempt>[0];
  isU2F: boolean;
};

type Props = {
  auth2faType?: Auth2faType;
  onChangePass(oldPass: string, newPass: string, token: string): Promise<any>;
  onChangePassWithU2f(oldPass: string, newPass: string): Promise<any>;
};

export default FormPassword;
