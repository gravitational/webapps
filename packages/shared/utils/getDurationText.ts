import { pluralize } from 'teleport/lib/util';

export function getDurationText(hrs: number, mins: number, secs: number) {
  if (!hrs && !mins) {
    return `${secs} secs`;
  }

  const hrText = pluralize(hrs, 'hr');
  const minText = pluralize(mins, 'min');

  if (!hrs) {
    return `${mins} ${minText}`;
  }

  if (hrs && !mins) {
    return `${hrs} ${hrText}`;
  }

  return `${hrs} ${hrText} and ${mins} ${minText}`;
}
