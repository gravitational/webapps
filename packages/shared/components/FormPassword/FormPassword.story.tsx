/*
Copyright 2019-2022 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';

import FormPassword from './FormPassword';

export default {
  title: 'Shared/FormPassword',
};

export const Off = () => <FormPassword {...props} />;

export const OTP = () => <FormPassword auth2faType="otp" {...props} />;

export const Webauthn = () => (
  <FormPassword auth2faType="webauthn" {...props} />
);

export const On = () => (
  <FormPassword auth2faType="on" {...props} preferredMfaType="webauthn" />
);

export const Optional = () => (
  <FormPassword auth2faType="optional" {...props} preferredMfaType="webauthn" />
);

const props = {
  onChangePass: () => Promise.resolve(),
  onChangePassWithWebauthn: () => Promise.reject(new Error('server error')),
};
