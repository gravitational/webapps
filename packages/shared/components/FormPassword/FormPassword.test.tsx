/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import FormPassword from './FormPassword';
import { On, Optional } from './FormPassword.story';
import { render, fireEvent, wait, screen } from 'design/utils/testing';

jest.mock('../../libs/logger', () => {
  const mockLogger = {
    error: jest.fn(),
  };

  return {
    create: () => mockLogger,
  };
});

const placeholdCurrPass = /^password$/i;
const placeholdNewPass = /new password/i;
const placeholdConfirm = /confirm password/i;

const btnSubmitText = /update password/i;

const inputValText = 'aaaaaa';
const inputVal = { target: { value: inputValText } };

test('input validation error states', async () => {
  const onChangePass = jest.fn().mockResolvedValue(null);
  const onChangePassWithU2f = jest.fn().mockResolvedValue(null);

  const { getByText } = render(
    <FormPassword
      auth2faType={'otp'}
      onChangePass={onChangePass}
      onChangePassWithU2f={onChangePassWithU2f}
    />
  );

  // test input validation error states
  await wait(() => fireEvent.click(getByText(btnSubmitText)));
  expect(onChangePass).not.toHaveBeenCalled();
  expect(onChangePassWithU2f).not.toHaveBeenCalled();

  expect(getByText(/current password is required/i)).toBeInTheDocument();
  expect(getByText(/enter at least 6 characters/i)).toBeInTheDocument();
  expect(getByText(/please confirm your password/i)).toBeInTheDocument();
  expect(getByText(/token is required/i)).toBeInTheDocument();
});

test('prop auth2faType: disabled', async () => {
  const onChangePass = jest.fn().mockResolvedValue(null);
  const onChangePassWithU2f = jest.fn().mockResolvedValue(null);

  const { getByText, getByPlaceholderText } = render(
    <FormPassword
      auth2faType="off"
      onChangePass={onChangePass}
      onChangePassWithU2f={onChangePassWithU2f}
    />
  );

  // fill out form
  fireEvent.change(getByPlaceholderText(placeholdCurrPass), inputVal);
  fireEvent.change(getByPlaceholderText(placeholdNewPass), inputVal);
  fireEvent.change(getByPlaceholderText(placeholdConfirm), inputVal);

  // test the correct cb is called when submitting
  await wait(() => fireEvent.click(getByText(btnSubmitText)));
  expect(onChangePass).toHaveBeenCalledWith(inputValText, inputValText, '');
  expect(onChangePassWithU2f).not.toHaveBeenCalled();

  // test rendering of status message after submit
  expect(getByText(/your password has been changed!/i)).toBeInTheDocument();

  // test clearing of form values after submit
  expect(getByPlaceholderText(placeholdCurrPass)).toHaveAttribute('value', '');
  expect(getByPlaceholderText(placeholdNewPass)).toHaveAttribute('value', '');
  expect(getByPlaceholderText(placeholdConfirm)).toHaveAttribute('value', '');
});

test('prop auth2faType: OTP form', async () => {
  const onChangePass = jest.fn().mockResolvedValue(null);
  const onChangePassWithU2f = jest.fn().mockResolvedValue(null);

  const { getByText, getByPlaceholderText } = render(
    <FormPassword
      auth2faType="otp"
      onChangePass={onChangePass}
      onChangePassWithU2f={onChangePassWithU2f}
    />
  );

  // test input validation error state
  await wait(() => fireEvent.click(getByText(btnSubmitText)));

  // fill out form
  fireEvent.change(getByPlaceholderText(placeholdCurrPass), inputVal);
  fireEvent.change(getByPlaceholderText(placeholdNewPass), inputVal);
  fireEvent.change(getByPlaceholderText(placeholdConfirm), inputVal);
  fireEvent.change(getByPlaceholderText(/otp token/i), inputVal);

  // test the correct cb is called when submitting
  await wait(() => fireEvent.click(getByText(btnSubmitText)));
  expect(onChangePass).toHaveBeenCalledWith(
    inputValText,
    inputValText,
    inputValText
  );
  expect(onChangePassWithU2f).not.toHaveBeenCalled();

  // test clearing of form values after submit
  expect(getByPlaceholderText(placeholdCurrPass)).toHaveAttribute('value', '');
  expect(getByPlaceholderText(placeholdNewPass)).toHaveAttribute('value', '');
  expect(getByPlaceholderText(placeholdConfirm)).toHaveAttribute('value', '');
  expect(getByPlaceholderText(/otp token/i)).toHaveAttribute('value', '');
});

test('prop auth2faType: U2f form with mocked error', async () => {
  const onChangePass = jest.fn().mockResolvedValue(null);
  const onChangePassWithU2f = jest.fn().mockRejectedValue(new Error('errMsg'));

  const { getByText, getByPlaceholderText } = render(
    <FormPassword
      auth2faType={'u2f'}
      onChangePass={onChangePass}
      onChangePassWithU2f={onChangePassWithU2f}
    />
  );

  // fill out form
  fireEvent.change(getByPlaceholderText(placeholdCurrPass), inputVal);
  fireEvent.change(getByPlaceholderText(placeholdNewPass), inputVal);
  fireEvent.change(getByPlaceholderText(placeholdConfirm), inputVal);

  // test U2F status message

  await wait(() => {
    fireEvent.click(getByText(btnSubmitText));
    const statusMsg = getByText(
      /Insert your U2F key and press the button on the key/i
    );
    expect(statusMsg).toBeInTheDocument();
  });

  // test correct cb is called
  expect(onChangePass).not.toHaveBeenCalled();
  expect(onChangePassWithU2f).toHaveBeenCalledTimes(1);

  // test rendering of status message after submit
  expect(getByText(/errMsg/i)).toBeInTheDocument();

  // test forms are NOT cleared with processing errors
  expect(getByPlaceholderText(placeholdCurrPass)).not.toHaveAttribute(
    'value',
    ''
  );
  expect(getByPlaceholderText(placeholdNewPass)).not.toHaveAttribute(
    'value',
    ''
  );
  expect(getByPlaceholderText(placeholdConfirm)).not.toHaveAttribute(
    'value',
    ''
  );
});

test('auth2faType "optional", has correct select options', async () => {
  render(<Optional />);

  // default mfa dropdown is set to none
  expect(screen.getByTestId('mfa-select').textContent).toMatch(/none/i);

  // select u2f from mfa dropdown
  let selectEl = screen.getByTestId('mfa-select').querySelector('input');
  fireEvent.focus(selectEl);
  fireEvent.keyDown(selectEl, { key: 'ArrowDown', keyCode: 40 });
  fireEvent.click(screen.getByText(/u2f/i));
  screen.getByText(/u2f/i);

  // select totp from mfa dropdown
  selectEl = screen.getByTestId('mfa-select').querySelector('input');
  fireEvent.focus(selectEl);
  fireEvent.keyDown(selectEl, { key: 'ArrowDown', keyCode: 40 });
  fireEvent.click(screen.getByText(/totp/i));
  screen.getByText(/totp/i);

  // test totp requirement
  fireEvent.click(screen.getByText(/update password/i));
  screen.getByText(/token is require/i);
});

test('auth2faType "on", has correct select options', async () => {
  render(<On />);

  // default mfa dropdown is set to u2f
  expect(screen.getByTestId('mfa-select').textContent).toMatch(/u2f/i);

  // select totp from mfa dropdown
  const selectEl = screen.getByTestId('mfa-select').querySelector('input');
  fireEvent.focus(selectEl);
  fireEvent.keyDown(selectEl, { key: 'ArrowDown', keyCode: 40 });
  fireEvent.click(screen.getByText(/totp/i));
  screen.getByText(/totp/i);

  // none type is not part of mfa dropdown
  fireEvent.focus(selectEl);
  fireEvent.keyDown(selectEl, { key: 'ArrowDown', keyCode: 40 });
  expect(screen.queryAllByText(/none/i)).toHaveLength(0);
});
