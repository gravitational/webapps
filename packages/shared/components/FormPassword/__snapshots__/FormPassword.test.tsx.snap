// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth2faType "on" should render form with hardware key as first option in dropdown 1`] = `
.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c5 {
  box-sizing: border-box;
  width: 50%;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 0px;
  margin-right: 16px;
}

.c9 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 40px;
  font-size: 12px;
  padding: 0px 40px;
  margin-top: 32px;
  width: 100%;
}

.c9:active {
  opacity: 0.56;
}

.c9:hover,
.c9:focus {
  background: #651FFF;
}

.c9:active {
  background: #354AA4;
}

.c9:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c0 {
  box-sizing: border-box;
  padding: 40px;
  width: 456px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c3 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
  margin-top: 4px;
}

.c3::-ms-clear {
  display: none;
}

.c3::placeholder {
  opacity: 0.4;
}

.c3:read-only {
  cursor: not-allowed;
}

.c2 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 0px;
}

.c7 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c4 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: flex-end;
}

.c8 .react-select-container {
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  border: none;
  display: block;
  font-size: 14px;
  outline: none;
  width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #ffffff;
  margin-bottom: 0px;
  border-radius: 4px;
}

.c8 .react-select__control,
.c8 .react-select__control--is-focused {
  min-height: 40px;
  height: 40px;
  background-color: transparent;
  border-color: transparent;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: none;
}

.c8 .react-select__control:hover,
.c8 .react-select__control--is-focused:hover {
  border-color: transparent;
  cursor: pointer;
}

.c8 .react-select__option:hover {
  cursor: pointer;
  background-color: #eceff1;
}

.c8 .react-select__option--is-focused {
  background-color: #eceff1;
}

.c8 .react-select__option--is-selected {
  background-color: #cfd8dc;
  color: inherit;
}

.c8 .react-select__option--is-selected:hover {
  background-color: #cfd8dc;
}

.c8 .react-select__menu {
  margin-top: 0px;
}

.c8 .react-select__indicator-separator {
  display: none;
}

.c8 .react-select__loading-indicator {
  display: none;
}

.c8 .react-select--is-disabled .react-select__single-value,
.c8 .react-select--is-disabled .react-select__placeholder {
  color: rgba(0,0,0,0.24);
}

.c8 .react-select--is-disabled .react-select__indicator {
  color: rgba(0,0,0,0.14);
}

<div>
  <form
    class="c0"
    width="456px"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        font-size="0"
      >
        Current Password
        <input
          autocomplete="off"
          class="c3"
          color="text.onLight"
          inputmode="text"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
        data-testid="mfa-select"
        width="50%"
      >
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
            for="select"
          >
            Two-factor type
          </label>
          <div
            class="c8"
          >
            <div
              class="react-select-container css-2b097c-container"
            >
              <div
                class="react-select__control css-yk16xz-control"
              >
                <div
                  class="react-select__value-container react-select__value-container--has-value css-g1d714-ValueContainer"
                >
                  <div
                    class="react-select__single-value css-1uccc91-singleValue"
                  >
                    Hardware Key
                  </div>
                  <input
                    aria-autocomplete="list"
                    class="css-62g3xt-dummyInput"
                    id="select"
                    readonly=""
                    tabindex="0"
                    value=""
                  />
                </div>
                <div
                  class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c5"
        width="50%"
      />
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        font-size="0"
      >
        New Password
        <input
          autocomplete="off"
          class="c3"
          color="text.onLight"
          inputmode="text"
          placeholder="New Password"
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        font-size="0"
      >
        Confirm Password
        <input
          autocomplete="off"
          class="c3"
          color="text.onLight"
          inputmode="text"
          placeholder="Confirm Password"
          type="password"
          value=""
        />
      </label>
    </div>
    <button
      class="c9"
      kind="primary"
    >
      Update Password
    </button>
  </form>
</div>
`;

exports[`auth2faType "optional" should render form with hardware key as first option in dropdown 1`] = `
.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c5 {
  box-sizing: border-box;
  width: 50%;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 0px;
  margin-right: 16px;
}

.c9 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 40px;
  font-size: 12px;
  padding: 0px 40px;
  margin-top: 32px;
  width: 100%;
}

.c9:active {
  opacity: 0.56;
}

.c9:hover,
.c9:focus {
  background: #651FFF;
}

.c9:active {
  background: #354AA4;
}

.c9:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c0 {
  box-sizing: border-box;
  padding: 40px;
  width: 456px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c3 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
  margin-top: 4px;
}

.c3::-ms-clear {
  display: none;
}

.c3::placeholder {
  opacity: 0.4;
}

.c3:read-only {
  cursor: not-allowed;
}

.c2 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 0px;
}

.c7 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c4 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: flex-end;
}

.c8 .react-select-container {
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  border: none;
  display: block;
  font-size: 14px;
  outline: none;
  width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #ffffff;
  margin-bottom: 0px;
  border-radius: 4px;
}

.c8 .react-select__control,
.c8 .react-select__control--is-focused {
  min-height: 40px;
  height: 40px;
  background-color: transparent;
  border-color: transparent;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: none;
}

.c8 .react-select__control:hover,
.c8 .react-select__control--is-focused:hover {
  border-color: transparent;
  cursor: pointer;
}

.c8 .react-select__option:hover {
  cursor: pointer;
  background-color: #eceff1;
}

.c8 .react-select__option--is-focused {
  background-color: #eceff1;
}

.c8 .react-select__option--is-selected {
  background-color: #cfd8dc;
  color: inherit;
}

.c8 .react-select__option--is-selected:hover {
  background-color: #cfd8dc;
}

.c8 .react-select__menu {
  margin-top: 0px;
}

.c8 .react-select__indicator-separator {
  display: none;
}

.c8 .react-select__loading-indicator {
  display: none;
}

.c8 .react-select--is-disabled .react-select__single-value,
.c8 .react-select--is-disabled .react-select__placeholder {
  color: rgba(0,0,0,0.24);
}

.c8 .react-select--is-disabled .react-select__indicator {
  color: rgba(0,0,0,0.14);
}

<div>
  <form
    class="c0"
    width="456px"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
        font-size="0"
      >
        Current Password
        <input
          autocomplete="off"
          class="c3"
          color="text.onLight"
          inputmode="text"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
        data-testid="mfa-select"
        width="50%"
      >
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
            for="select"
          >
            Two-factor type
          </label>
          <div
            class="c8"
          >
            <div
              class="react-select-container css-2b097c-container"
            >
              <div
                class="react-select__control css-yk16xz-control"
              >
                <div
                  class="react-select__value-container react-select__value-container--has-value css-g1d714-ValueContainer"
                >
                  <div
                    class="react-select__single-value css-1uccc91-singleValue"
                  >
                    Hardware Key
                  </div>
                  <input
                    aria-autocomplete="list"
                    class="css-62g3xt-dummyInput"
                    id="select"
                    readonly=""
                    tabindex="0"
                    value=""
                  />
                </div>
                <div
                  class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-6q0nyr-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c5"
        width="50%"
      />
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        font-size="0"
      >
        New Password
        <input
          autocomplete="off"
          class="c3"
          color="text.onLight"
          inputmode="text"
          placeholder="New Password"
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        font-size="0"
      >
        Confirm Password
        <input
          autocomplete="off"
          class="c3"
          color="text.onLight"
          inputmode="text"
          placeholder="Confirm Password"
          type="password"
          value=""
        />
      </label>
    </div>
    <button
      class="c9"
      kind="primary"
    >
      Update Password
    </button>
  </form>
</div>
`;
